### PR TITLE
Compute rational signature_tuple via leading principal minors

### PR DESCRIPTION
**Pull request by simonbrandhorst**
## Summary
- `signature_tuple(q::QuadSpace{QQField,QQMatrix})` now determines the signature from the sequence of leading principal minors $D_0=1, D_1, \dots, D_m$ of a Gram matrix of the non-degenerate part of `q`, using Jacobi's theorem (the number of sign changes in this sequence equals the number of negative eigenvalues). This avoids computing an explicit basis-change matrix.
- Each minor is checked as soon as it is computed, and the function bails out to the fallback the moment one is zero, instead of computing the full sequence of (increasingly expensive) minors first and only then checking whether any of them vanished.
- When the Gram matrix is already non-degenerate (kernel trivial, the common case), the basis-completion step and the two matrix multiplications it would otherwise take to restrict to the non-degenerate part are skipped entirely, since they would be a no-op (identity) in that case.
- When a leading principal minor vanishes (e.g. a hyperbolic plane sitting on the diagonal, where Jacobi's theorem does not directly apply), the function falls back to the previous exact Gram-Schmidt diagonalization, which handles this case via pivoting.
- The result is still exact (`QQFieldElem` arithmetic throughout), so `signature_tuple` remains provably correct.

## Benchmarks

With #2345 merged, `diagonal`/`_gram_schmidt` itself got much faster, so it was worth re-checking whether this PR is still worthwhile on top of that. Compared `signature_tuple` from this PR against always calling `diagonal(q)` (i.e. the pre-this-PR behavior, now benefiting from #2345), on top of current `master`:

**Generic full-rank matrices (no fallback triggered):**

| n | diag (ms) | minors (ms) | speedup | diag alloc | minors alloc | alloc ratio |
|---|---|---|---|---|---|---|
| 5 | 0.018 | 0.004 | 4.54x | 25,951 | 5,106 | 5.08x |
| 10 | 0.121 | 0.018 | 6.71x | 163,859 | 18,912 | 8.66x |
| 20 | 7.69 (*) | 0.175 | 43.9x (*) | 1,303,573 | 91,434 | 14.26x |
| 40 | 38.27 | 1.86 | 20.55x | 12,344,300 | 1,461,624 | 8.45x |
| 80 | 769.32 | 40.84 | 18.84x | 107,604,624 | 17,098,267 | 6.29x |
| 120 | 3461.02 | 381.00 | 9.08x | 300,421,408 | 64,551,560 | 4.65x |

(*) the n=20 "diag" timing looks like a noisy outlier relative to its neighbors; treat that one speedup figure loosely, the rest of the trend is consistent. Still a clear win across the board, roughly 5x-20x depending on size, with allocations lower by a similar margin.

**Matrices with a zero leading diagonal entry** (e.g. a hyperbolic plane in the standard basis, common for real lattice Gram matrices), which force the fallback path — across the three iterations of this PR:

| case | original (compute all minors first) | + early bail-out | + skip basis completion when already non-degenerate |
|---|---|---|---|
| n=10 | 0.72x | 0.81x | 0.95x |
| n=20 | 0.71x | 0.81x | 0.95x |
| n=40 | 0.37x | 0.81x | 0.91x |

Originally, computing the full sequence of minors before discarding them made this case increasingly slower than the plain `diagonal(q)` fallback as `n` grew (down to 0.37x at n=40, unbounded further out). The early bail-out capped that at a flat ~19% overhead regardless of `n`. Since these particular matrices are already full rank (trivial kernel), the basis-completion step was pure unnecessary overhead here too; skipping it brings the remaining gap down to ~5-9%, essentially just the cost of the one trivial determinant before falling back.

## Test plan
- [x] Added a `signature_tuple via leading principal minors` testset in `test/QuadForm/Quad/Spaces.jl` covering: empty matrix, isotropic/definite 1x1, hyperbolic plane (forces the fallback path), rank-1 degenerate forms, and a mixed degenerate/indefinite 4x4 example.
- [x] Cross-checked against an explicit Gram-Schmidt diagonalization on 131 random/edge-case matrices (including degenerate and hyperbolic-plane cases that trigger the fallback), re-run after each change in this PR.
- [x] `test/QuadForm/Quad/Spaces.jl` (2632 tests) passes.
- [x] `test/QuadForm/Quad/ZGenus.jl` (1332 tests, 1 unrelated harness-only error from running the file standalone) passes.
- [x] `test/QuadForm/Quad/ZLattices.jl` (1371 tests, 1 pre-existing unrelated broken test) passes.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: test/QuadForm/Quad/Spaces.jl [hunk 1]
```diff
@@ -5,14 +5,45 @@
...
+        g = A + transpose(A)
+        V = quadratic_space(QQ, g)
+        D = diagonal(V)
+        pos = count(d -> d > 0, D)
+        zer = count(d -> d == 0, D)
+        neg = count(d -> d < 0, D)
+        @test Hecke.signature_tuple(V) == (pos, zer, neg)
+      end
+    end
+  end
```
Comment (simonbrandhorst) [inline test/QuadForm/Quad/Spaces.jl]:
```suggestion
```

**Review by simonbrandhorst — COMMENTED** [review]
AI changes look good to me.

File: src/QuadForm/Quad/Spaces.jl [hunk 2]
```diff
@@ -2138,13 +2117,47 @@ end
     signature_tuple(q::QuadraticSpace{QQField,QQMatrix) ->Tuple{Int,Int,Int}
 
 Return the number of (positive, zero, negative) inertia of this rational quadratic space.
+
+This is computed from the sequence of leading principal minors of a Gram
+matrix of the non-degenerate part of `q` via Jacobi's theorem, which is
+usually faster than computing an explicit diagonalization.
```
Comment (simonbrandhorst) [inline src/QuadForm/Quad/Spaces.jl]:
```suggestion
```

File: src/QuadForm/Quad/Spaces.jl [hunk 3]
```diff
@@ -2140,11 +2119,41 @@ end
...
+    # are nonzero, the number of negative eigenvalues equals the number of
+    # sign changes in the sequence D_0, D_1, ..., D_m.
+    neg = count(r -> (minors[r] > 0) != (minors[r + 1] > 0), 1:m)
+    pos = m - neg
+    return (pos, k, neg)
+  end
+
+  # Jacobi's theorem does not apply directly when a leading principal minor
+  # vanishes (e.g. a hyperbolic plane on the diagonal); fall back to an
+  # explicit diagonalization, which handles this case via pivoting.
```
Comment (simonbrandhorst) [inline src/QuadForm/Quad/Spaces.jl]:
```suggestion
  # explicit diagonalization
```

File: test/QuadForm/Quad/Spaces.jl [hunk 4]
```diff
@@ -5,14 +5,31 @@
...
+    # hyperbolic plane: the leading principal minor D_1 vanishes, forcing the
+    # fallback diagonalization
+    @test Hecke.signature_tuple(quadratic_space(QQ, QQ[0 1; 1 0])) == (1, 0, 1)
+    # rank-1 degenerate form
+    @test Hecke.signature_tuple(quadratic_space(QQ, QQ[1 1; 1 1])) == (1, 1, 0)
+    @test Hecke.signature_tuple(quadratic_space(QQ, QQ[2 1; 1 2])) == (2, 0, 0)
+    @test Hecke.signature_tuple(quadratic_space(QQ, QQ[-2 1; 1 -2])) == (0, 0, 2)
+    # hyperbolic plane plus a definite direction, with a degenerate direction
+    g = matrix(QQ, [0 1 0 0; 1 0 0 0; 0 0 -3 0; 0 0 0 0])
+    @test Hecke.signature_tuple(quadratic_space(QQ, g)) == (1, 1, 2)
```
Comment (simonbrandhorst) [inline test/QuadForm/Quad/Spaces.jl:25]:
```suggestion
    @test Hecke.signature_tuple(quadratic_space(QQ, g)) == (1, 1, 2)
  end
```

**Comment by simonbrandhorst** [comment]
Let's wait for #2345 and then measure whether this is still an improvement.

**Comment by thofma** [comment]
https://github.com/thofma/Hecke.jl/pull/2345 has landed

**Comment by simonbrandhorst** [comment]
Ready to go from my side. More could still be done but I guess that would be over engineering it.
